libsql/internal/http/http.go

```diff
@@ -90,25 +119,124 @@ func callSqld(ctx context.Context, url string, sql string, sqlParams params) (*r
+	for idx := range args {
+		sortedArgs = append(sortedArgs, &args[idx])
+	}
+	sort.Slice(sortedArgs, func(i, j int) bool {
+		return sortedArgs[i].Ordinal < sortedArgs[j].Ordinal
+	})
+
+	params := parameters{paramsType: getParamType(sortedArgs[0])}
+	for _, arg := range sortedArgs {
+		if params.paramsType != getParamType(arg) {
+			return parameters{}, fmt.Errorf("driver does not accept positional and named parameters at the same time")
+		}
+		params.paramsValues = append(params.paramsValues, struct {
+			name  string
+			value any
+		}{name: arg.Name, value: arg.Value})
+	}
+	return params, nil
+}
+
+func getParamType(arg *driver.NamedValue) paramsType {
+	if arg.Name == "" {
+		return positionalParameters
+	}
+	return namedParameters
+}
+
+func generatePostBody(sql string, sqlParams parameters) (*postBody, error) {
+	stmts, err := sqlparser.SplitStatementToPieces(sql)
+	if err != nil {
+		return nil, err
+	}
+
+	postBody := postBody{}
+
+	totalParametersAlreadyUsed := 0
+	for _, stmt := range stmts {
+		stmtParameters, err := generateStatementParameters(stmt, sqlParams, totalParametersAlreadyUsed)
+		if err != nil {
+			return nil, fmt.Errorf("fail to generate statement parameter. statement: %s. error: %v", stmt, err)
+		}
+		postBody.Statements = append(postBody.Statements, statement{stmt, stmtParameters})
+		totalParametersAlreadyUsed += len(stmtParameters.paramsValues)
+	}
+
+	return &postBody, nil
+
+}
+
+func generateStatementParameters(stmt string, queryParams parameters, positionalParametersOffset int) (parameters, error) {
+	parsedStmt, err := sqlparser.Parse(stmt)
+	if err != nil {
+		return parameters{}, err
+	}
+
+	stmtParametersNames := make([]string, 0)
+	err = sqlparser.Walk(func(node sqlparser.SQLNode) (bool, error) {
+		if sqlVal, ok := node.(*sqlparser.SQLVal); ok && sqlVal.Type == sqlparser.ValArg {
+			paramName, err := removeParamPrefix(string(sqlVal.Val))
+			if err != nil {
+				return false, err
+			}
+			stmtParametersNames = append(stmtParametersNames, paramName)
+		}
+		return true, nil
+	}, parsedStmt)
+	if err != nil {
+		return parameters{}, err
+	}
+
+	stmtParams := parameters{paramsType: queryParams.paramsType}
+
+	switch queryParams.paramsType {
+	case positionalParameters:
+		if positionalParametersOffset+len(stmtParametersNames) >= len(queryParams.paramsValues) {
+			return parameters{}, fmt.Errorf("missing positional parameters")
+		}
+		stmtParams.paramsValues = queryParams.paramsValues[positionalParametersOffset : positionalParametersOffset+len(stmtParametersNames)]
+	case namedParameters:
+		stmtParametersNeeded := make(map[string]bool)
+		for _, stmtParametersName := range stmtParametersNames {
+			stmtParametersNeeded[stmtParametersName] = true
+		}
+		for _, queryParamsValue := range queryParams.paramsValues {
+			if stmtParametersNeeded[queryParamsValue.name] {
+				stmtParams.paramsValues = append(stmtParams.paramsValues, queryParamsValue)
+				stmtParametersNeeded[queryParamsValue.name] = false
```

Comment:
Why do we have to clean this up? Do we expect the same param to appear twice?

Reply:
queryParams.paramsValues is just a convertion from what `[]driver.Namedvalue` sent to us. Since we don't have control over that, I decided to go safe here

Reply:
Fair enough! I think it's a good idea.